xresolver/types.go

```diff
@@ -0,0 +1,181 @@
+package xresolver
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/url"
+	"strconv"
+	"sync"
+)
+
+type Lookup interface {
+	// LookupIPAddr looks up host using the local resolver. It returns a slice of that host's IPv4 and IPv6 addresses.
+	LookupRoutes(ctx context.Context, host string) ([]Route, error)
+}
+
+type Dial interface {
+	// DialContext connects to the address on the named network using the provided context.
+	DialContext(ctx context.Context, network, addr string) (con net.Conn, err error)
+}
+
+type ConnCreation interface {
+	Dial(network, address string) (net.Conn, error)
+}
+
+// Resolver represents how to generate the address and how to create the connection
+type Resolver interface {
+	Dial
+
+	// Add adds the resolver to the methods of creating the IPv4 and IPv6 addresses
+	Add(r Lookup) error
+
+	// Remove removes the resolver to the methods of creating the IPv4 and IPv6 addresses
+	Remove(r Lookup) error
+}
+
+type Route struct {
+	Scheme string
+	Host   string
+	Port   int
+}
+
+func CreateRoute(route string) (Route, error) {
+	path, err := url.Parse(route)
+	if err != nil {
+		return Route{}, err
+	}
+	port, err := strconv.Atoi(path.Port())
+	return Route{
+		Scheme: path.Scheme,
+		Host:   path.Hostname(),
+		Port:   port,
+	}, err
+}
+
+func (r Route) String() string {
+	str := r.Scheme + "://" + r.Host
+	if r.Port != 0 {
+		return str + ":" + strconv.Itoa(r.Port)
+	}
+	return str
+}
+
+type orderedRoute struct {
+	route Route
+	index int
+}
+
+type RoundRobin struct {
+	lock   sync.RWMutex
+	routes map[string]*orderedRoute
+}
+
+func NewRoundRobinBalancer() *RoundRobin {
+	return &RoundRobin{
+		routes: make(map[string]*orderedRoute),
+	}
+}
+
+func (robin *RoundRobin) Add(route Route) error {
+	// check if exist
+	robin.lock.RLock()
```

Comment:
You should probably avoid the two `RUnlock` calls here, as this is a future bug waiting to happen.  This works better:

```
robin.lock.RLock()
_, found := robin.routes[route.String()]
robin.lock.RUnlock()
if found {
    return errors.New("addr already in rotation")
}
```

That's a lot easier to reason about.  And future bug fixes have a much lower risk of messing up the lock/unlock code.

Reply:
I would a channel be a better solution?

Reply:
no, there's no reason to force a context switch here.  a mutex is far better (or, an atomic variable).